### PR TITLE
loginController now emitting failedLogin hook after failed login attempts

### PR DESCRIPTION
**Pull request by karakayasemi**
## Description
There is an error in the logic of emitting preLogin signal in the LoginController. UserSession's login function is checking password and emitting preLogin signal, but loginController is either checking password before UserSession and returning view response without hook signal when failed login attempts.

With this change, unnecessary checkPassword functions are removed. Code simplified and now we are emitting preLogin hook also before failed login attempts.
## Motivation and Context
preLogin hook is not emitting on failed login attempts. To resolve following issue 
https://github.com/owncloud/security/issues/1

## How Has This Been Tested?
Unit tests are working

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/Controller/LoginController.php [hunk 1]
```diff
@@ -182,27 +182,22 @@ public function showLoginForm($user, $redirect_url, $remember_login) {
...
-		/* @var $loginResult IUser */
-		$loginResult = $this->userManager->checkPassword($user, $password);
-		if ($loginResult === false) {
-			$users = $this->userManager->getByEmail($user);
-			// we only allow login by email if unique
-			if (count($users) === 1) {
-				$user = $users[0]->getUID();
-				$loginResult = $this->userManager->checkPassword($user, $password);
-			}
+		$emailUsers = $this->userManager->getByEmail($user);
```
Comment (DeepDiver1975) [inline core/Controller/LoginController.php]: this totally removes the capability to login with the login name - right?

File: core/Controller/LoginController.php [hunk 2]
```diff
@@ -182,27 +182,22 @@ public function showLoginForm($user, $redirect_url, $remember_login) {
...
+		if(count($emailUsers) === 1) {
+			$user = $emailUsers[0]->getUID();
 		}
-		if ($loginResult === false) {
+		if($this->userSession->login($user, $password) !== true) {
```
Comment (DeepDiver1975) [inline core/Controller/LoginController.php]: so login is enough - we not longer need to call checkPassword? looks strange ....

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by DeepDiver1975** [comment]
an please squash your commits - thx

File: core/Controller/LoginController.php (hunk 2 above)
Comment (karakayasemi) [inline core/Controller/LoginController.php]: login is also using userManager's checkPassword inside of it with emitting hook signal before and after checkPassword. IMHO we can trust login.

File: core/Controller/LoginController.php (hunk 1 above)
Comment (karakayasemi) [inline core/Controller/LoginController.php]: If someone is trying to login with uid, getByEmail does not return anything. If a user's uid is the same as another user's email, it creates a problem.

**Comment by karakayasemi** [comment]
I messed up trying to squash. Trying to resolve, but i am not a git expert.

**Comment by DeepDiver1975** [comment]
We miss a login failed hook - right?

**Comment by karakayasemi** [comment]
we can add login failed hook to userSession's login method. However, I am not sure, where should we emit login failed signal also? For now, this change allows to use preLogin hook and it is enough to start security app. Should I work on login failed hook?
And also should I revert e-mail login part of the code? IMHO, it is simplifying logic and does not create any problem unless if a user's uid is the same as another user's email.

**Comment by DeepDiver1975** [comment]
> And also should I revert e-mail login part of the code? IMHO, it is simplifying logic and does not create any problem unless if a user's uid is the same as another user's email.

yes please - we shall keep the behavior

**Comment by DeepDiver1975** [comment]
> we can add login failed hook to userSession's login method. However, I am not sure, where should we emit login failed signal also? For now, this change allows to use preLogin hook and it is enough to start security app. Should I work on login failed hook?

we need that hook to log a failed login - or am I missing something?

**Comment by karakayasemi** [comment]
> we need that hook to log a failed login - or am I missing something?

IMHO, Login hook signal should emit from UserSession's login method. Since failed login check is done by userManager's checkPassword method in our current password correction logic, we can not emit any hook signal from userSessions login method in failed login attempts. Because of that, I removed checkPassword methods and used userSession's login function.
I will try to make changes you said in today.

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -464,7 +466,7 @@ private function loginWithPassword($uid, $password) {
 		$this->manager->emit('\OC\User', 'preLogin', [$uid, $password]);
 		$user = $this->manager->checkPassword($uid, $password);
 		if ($user === false) {
-			// Password check failed
+			$this->manager->emit('\OC\User', 'failedLogin', [$uid, $password]);
```
Comment (DeepDiver1975) [inline lib/private/User/Session.php]: shall we really expose the password?

Comment (karakayasemi) [inline lib/private/User/Session.php]: I thought, maybe it can be useful for some warning like "it is your previous password".

Comment (karakayasemi) [inline lib/private/User/Session.php]: Should I change password exposition? Still, I think it might be useful in fact.

Comment (DeepDiver1975) [inline lib/private/User/Session.php]: I would not expose the password here - @Peter-Prochaska objections?

Comment (phil-davis) [inline lib/private/User/Session.php]: As a user, I might have accidentally put in a password for an account of mine on some other site, so would rather not have it displayed/logged... anywhere.

**Comment by karakayasemi** [comment]
Now, failed login hook is exposing the only username.

**Comment by PVince81** [comment]
@DeepDiver1975 is this now acceptable for merging ?

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by PVince81** [comment]
This introduced a regression which was now fixed in https://github.com/owncloud/core/pull/28450

**Pull request merged**